Video-Transformers-Kotlin/app/src/main/java/com/tokbox/sample/basicvideochat/MainActivity.kt

```diff
@@ -0,0 +1,325 @@
+            }
+        } else {
+            EasyPermissions.requestPermissions(
+                this,
+                getString(R.string.rationale_video_app),
+                PERMISSIONS_REQUEST_CODE,
+                *perms
+            )
+        }
+    }
+
+    /* Make a request for session data */
+    private fun getSession() {
+        Log.i(TAG, "getSession")
+
+        apiService?.session?.enqueue(object : Callback<GetSessionResponse?> {
+            override fun onResponse(call: Call<GetSessionResponse?>, response: Response<GetSessionResponse?>) {
+                response.body()?.also {
+                    initializeSession(it.apiKey, it.sessionId, it.token)
+                }
+            }
+
+            override fun onFailure(call: Call<GetSessionResponse?>, t: Throwable) {
+                throw RuntimeException(t.message)
+            }
+        })
+    }
+
+    private fun initializeSession(apiKey: String, sessionId: String, token: String) {
+        Log.i(TAG, "apiKey: $apiKey")
+        Log.i(TAG, "sessionId: $sessionId")
+        Log.i(TAG, "token: $token")
+
+        /*
+        The context used depends on the specific use case, but usually, it is desired for the session to
+        live outside of the Activity e.g: live between activities. For a production applications,
+        it's convenient to use Application context instead of Activity context.
+         */
+        session = Session.Builder(this, apiKey, sessionId).build().also {
+            it.setSessionListener(sessionListener)
+            it.connect(token)
+        }
+    }
+
+    private fun initRetrofit() {
+        val logging = HttpLoggingInterceptor()
+        logging.setLevel(HttpLoggingInterceptor.Level.BODY)
+        val client: OkHttpClient = OkHttpClient.Builder()
+            .addInterceptor(logging)
+            .build()
+
+        retrofit = Retrofit.Builder()
+            .baseUrl(ServerConfig.CHAT_SERVER_URL)
+            .addConverterFactory(MoshiConverterFactory.create())
+            .client(client)
+            .build().also {
+                apiService = it.create(APIService::class.java)
+            }
+    }
+
+    private fun finishWithMessage(message: String) {
+        Log.e(TAG, message)
+        Toast.makeText(this, message, Toast.LENGTH_LONG).show()
+        finish()
+    }
+
+    companion object {
+        private val TAG = MainActivity::class.java.simpleName
+        private const val PERMISSIONS_REQUEST_CODE = 124
+    }
+
+    class logoWatermark(private val resources: Resources) : CustomVideoTransformer {
+
+        // Get the image in bitmap format
+        private var image: Bitmap = BitmapFactory.decodeResource(resources, R.drawable.vonage_logo)
+
+        fun resizeImage(image: Bitmap?, width: Int, height: Int): Bitmap {
+            return Bitmap.createScaledBitmap(image!!, width, height, true)
+        }
+
+        override fun onTransform(frame: BaseVideoRenderer.Frame) {
```

Comment:
Is this invoked directly by some webrtc thread? If it is, can we make `image` immutable for better thread safety? In any case the call to `resizeImage` is done in every iteration.

Reply:
I can make image immutable and assign the` resizeImage ` return value to a new Bitmap variable

Reply:
It depends... actually do you think we need some kind of cache for performance improvement? Maybe we don't. Rescaling this image may be irrelevant in terms of performance compared to all the webrtc stuff.